### ISSUE TITLE
You use Diou loss/Ciou loss only on rcnn head or both rpn and rcnn head?

### ISSUE THREAD

**Issue by XianYuZhiWang** [Zzh-tju/DIoU#8]

**Comment by XianYuZhiWang** [comment]
I use ciou loss on cascade-rcnn, I use it on rpn head and three cascade head, but the MAP and AP.5 decreased. which part's head should i use ciou loss?  By the way, do you have any suggest about loss weight of ciou loss？

**Comment by Zzh-tju** [comment]
describe your Cascade rcnn's repo, mmdet or Detectron2? and the CIoU code you use. What's the regression loss weight you use?

**Comment by XianYuZhiWang** [comment]
I used mmdetection2, loss weight 1, I read you code , you used weight 12, but I only get 10 categories,  maybe not need 12.
'''python
def ciou_loss(pred, target, eps=1e-7):
    """
    Generalized Intersection over Union: A Metric and A Loss for
    Bounding Box Regression
    Args:
        pred (torch.Tensor): Predicted bboxes of format (x1, y1, x2, y2),
            shape (n, 4).
        target (torch.Tensor): Corresponding gt bboxes, shape (n, 4).
        eps (float): Eps to avoid log(0).

    Return:
        Tensor: Loss tensor.
    """
    # overlap
    lt = torch.max(pred[:, :2], target[:, :2])
    rb = torch.min(pred[:, 2:], target[:, 2:])
    wh = (rb - lt).clamp(min=0)
    overlap = wh[:, 0] * wh[:, 1]

    # union
    ap = (pred[:, 2] - pred[:, 0]) * (pred[:, 3] - pred[:, 1])
    ag = (target[:, 2] - target[:, 0]) * (target[:, 3] - target[:, 1])
    union = ap + ag - overlap + eps

    # IoU
    ious = overlap / union

    # wh
    w_pred = pred[:, 2] - pred[:, 0]
    h_pred = pred[:, 3] - pred[:, 1]
    w_gt = target[:, 2] - target[:, 0]
    h_gt = target[:, 3] - target[:, 1]

    # min point, max point
    xc1 = torch.min(pred[:, 0], target[:, 0])
    yc1 = torch.min(pred[:, 1], target[:, 1])
    xc2 = torch.max(pred[:, 2], target[:, 2])
    yc2 = torch.max(pred[:, 3], target[:, 3])

    # center
    x_p = (pred[:, 0] + pred[:, 2])/2
    y_p = (pred[:, 1] + pred[:, 3])/2
    x_g = (target[:, 0] + target[:, 2])/2
    y_g = (target[:, 1] + target[:, 3])/2

    c = ((xc2 - xc1) ** 2) + ((yc2 - yc1) ** 2) + eps
    d = ((x_p - x_g) ** 2) + ((y_p - y_g) ** 2)
    u = d / c

    v = (4 / (math.pi ** 2)) * torch.pow((torch.atan(w_gt/h_gt)-torch.atan(w_pred/h_pred)), 2)
    with torch.no_grad():
        S = 1 - ious
        alpha = v / (S + v)

    ciou = ious - (u + alpha * v)
    loss = 1-ciou

    return loss
'''

**Comment by Zzh-tju** [comment]
你得确定一下函数输入的box坐标是什么范围的

**Comment by XianYuZhiWang** [comment]
是确定box坐标是否根据输入图片长宽归一化吗？

**Comment by Zzh-tju** [comment]
你打印出来我看看，输入的boxes

**Comment by XianYuZhiWang** [comment]
![Screenshot from 2020-07-09 18-36-19](https://user-images.githubusercontent.com/44389654/87029767-2dfddb00-c213-11ea-9d3c-41e1f3dab418.png)

**Comment by Zzh-tju** [comment]
mmdet原始就是这样的吗？都是真实尺寸的？

**Comment by XianYuZhiWang** [comment]
是的，计算 ciou loss 需要除以输入长宽归一化吗？

**Comment by Zzh-tju** [comment]
那这样不行，现在boxes四个坐标相当于网络的输出节点，原始Smooth L1可以产生大梯度，而IoU loss产生不了大的梯度，等于没起到什么优化作用

**Comment by Zzh-tju** [comment]
而且我很好奇rcnn系列不是回归偏移量吗？偏移量都会除以anchor的宽度或高度，怎么会产生这种看似真实坐标的数字。即便如此，回归损失肯定很大，用Smooth L1会产生几十几百的损失，怎么与class loss平衡？

**Comment by XianYuZhiWang** [comment]
计算 Smooth L1 之前会有一个coder
    def bbox2delta(proposals, gt, means=(0., 0., 0., 0.), stds=(1., 1., 1., 1.)):
        """Compute deltas of proposals w.r.t. gt.
    
        We usually compute the deltas of x, y, w, h of proposals w.r.t ground
        truth bboxes to get regression target.
        This is the inverse function of `delta2bbox()`
    
        Args:
            proposals (Tensor): Boxes to be transformed, shape (N, ..., 4)
            gt (Tensor): Gt bboxes to be used as base, shape (N, ..., 4)
            means (Sequence[float]): Denormalizing means for delta coordinates
            stds (Sequence[float]): Denormalizing standard deviation for delta
                coordinates
    
        Returns:
            Tensor: deltas with shape (N, 4), where columns represent dx, dy,
                dw, dh.
    
        """
        assert proposals.size() == gt.size()
    
        proposals = proposals.float()
        gt = gt.float()
        px = (proposals[..., 0] + proposals[..., 2]) * 0.5
        py = (proposals[..., 1] + proposals[..., 3]) * 0.5
        pw = proposals[..., 2] - proposals[..., 0]
        ph = proposals[..., 3] - proposals[..., 1]
    
        gx = (gt[..., 0] + gt[..., 2]) * 0.5
        gy = (gt[..., 1] + gt[..., 3]) * 0.5
        gw = gt[..., 2] - gt[..., 0]
        gh = gt[..., 3] - gt[..., 1]
    
        dx = (gx - px) / pw
        dy = (gy - py) / ph
        dw = torch.log(gw / pw)
        dh = torch.log(gh / ph)
        deltas = torch.stack([dx, dy, dw, dh], dim=-1)
    
        means = deltas.new_tensor(means).unsqueeze(0)
        stds = deltas.new_tensor(stds).unsqueeze(0)
        deltas = deltas.sub_(means).div_(stds)
    
        return deltas

**Comment by Zzh-tju** [comment]
你是否用了这个函数？

**Comment by XianYuZhiWang** [comment]
mmdetection2 是有参数控制计算loss时是使用coder之后的delta还是原坐标bbox的，我设置的是使用bbox，所以没有在loss的方法里解码。如果计算ciou的时候使用的xyxy是真实坐标的话，ciou代码应该是没问题的，而且训练结果map从0.81下降到0.77，还算正常。我现在考虑的问题是，我在rpn和cascade三个头都用ciou，我是不是不该在rpn阶段用ciou，我看你源码好像rpn阶段是没有用的。另外就是损失的权重该怎么设置？是不是设置一个权重使得bbox loss和cls loss差不多就行了，我就得权重12的话在我数据及上可能有点大
